index.rst

```diff
@@ -1,29 +1,490 @@
+
+json schema was not designed for schema evolution, but Kafka does support it, as `described here <https://yokota.blog/2021/03/29/understanding-json-schema-compatibility/>`_.
+We tested schema evolution and found, as explained in that link, that setting additionalProperties=false allows one to add fields, but not delete them.
+Fortunately, setting additionalProperties=false is our preferred way to define schemas, because it detects typos in field names.
+However, the inability to delete fields is a drawback.
+
+Arrays
+------
+
+SAL arrays are all fixed-length.
+This is a significant issue, because neither Avro nor protobuf supports any kind of constraint on array length.
+json schema is the only choice that supports array length constraints, and we would have to test if Kafka enforces them.
+
+In order to constrain array lengths using Avro or Google protobuf we will have to manually check the data ourselves.
+This is practical as long as we retain additional information (beyond that used to generate the schema used by Kafka) when we parse the XML files that define our SAL interfaces.
+The Kafka version of ts_salobj checks array lengths.
+
+Scalar Data Types
+-----------------
+
+SAL supports the following scalar field types: boolean, byte/octet (uint8), short (int16), unsigned short (uint16), int/long (int32), unsigned int/long (uint32), long long (int64), unsigned long long (uint64), float, double, string (unicode).
+The slashes indicate pairs of names that mean the same thing, e.g. byte and octet both translate to uint8.
+Whichever Kafka schema we choose must also support these, using a larger type for integers and double for float, if necessary.
+Note that this is true whether or not we use Kafka for T&S communication, because we use Kafka to ingest T&S messages into the Engineering Facilities Database.
+
+Avro supports null, boolean, int (int3232), long (int64), float, double, bytes (sequence of 8-bit unsigned values), and string (unicode).
+This supports all the SAL data types except uint64, for which int64 is a poor representation.
+Fortunately uint64 fields are only used for a few fields for one SAL component: MTAOS (plus the Test component, which exercises all supported field types), and it is likely we can eliminate that usage.
+We must solve this, in order to properly ingest the data into the EFD.
+
+Protobuf supports supports bool, float, double, int32, int64, uint32, uint64 (and several variants of these with different encoding), float, double, bytes, string.
+This includes all of the SAL data types.
+
+json schema supports the following scalar types: null, boolean, integer (we have not found a specification for the size), number (float or integer), string.
+Note the weak support for numeric data types: only integer and float.
+
+Default Float Value
+-------------------
+
+The default value for float is a tricky question.
+NaN is most principled, but it has issues:
```

Comment:
Angelo proposed NaN and this was discussed at some length. I think it's a wonderful idea, but for the reasons detailed here I fear it might not be practical.

If Avro is updated to allow NaN in all languages then it is worth revisiting. But I think it greatly limits our options to migrate the software in the future. 0 is clearly the most compatible option.

Reply:
https://jira.lsstcorp.org/browse/RFC-763